codespan-reporting/src/term/utils.rs

```diff
@@ -0,0 +1,102 @@
+use std::ops::Range;
+/*
+pub(crate) fn closest_char_boundary(s: &str, byte_index: usize) -> usize {
+    let length = s.len();
+    if let Some(index) = (byte_index..=length)
+        .skip_while(|&i| !s.is_char_boundary(i))
+        .next()
+    {
+        index
+    } else {
+        length
+    }
+}
+
+pub(crate) fn closest_char_boundary_rev(s: &str, byte_index: usize) -> usize {
+    if let Some(index) = (0..=byte_index)
+        .rev()
+        .skip_while(|&i| !s.is_char_boundary(i))
+        .next()
+    {
+        index
+    } else {
+        0
+    }
+}
+*/
+
+pub(crate) fn closest_char_boundary(s: &str, byte_index: usize) -> usize {
+    let length = s.len();
+    for index in byte_index..=length {
+        if s.is_char_boundary(index) {
+            return index;
+        }
+    }
+    length
+}
+
+pub(crate) fn closest_char_boundary_rev(s: &str, byte_index: usize) -> usize {
+    for index in (0..=byte_index).rev() {
+        if s.is_char_boundary(index) {
+            return index;
+        }
+    }
+    0
+}
+
+pub(crate) fn slice_at_char_boundaries<'a>(s: &'a str, range: Range<usize>) -> &'a str {
+    let start = closest_char_boundary_rev(s, range.start);
+    let end = closest_char_boundary(s, range.end);
+    &s[start..end]
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::iter::repeat;
+
+    #[test]
+    fn test_boundary() {
+        let s = "🌞";
+        assert_eq!(closest_char_boundary(s, 0), 0);
+        assert_eq!(closest_char_boundary_rev(s, 0), 0);
+        for i in 1..s.len() {
+            assert_eq!(closest_char_boundary_rev(s, i), 0);
+            assert_eq!(closest_char_boundary(s, i), s.len());
+        }
+    }
+
+    #[test]
+    fn test_boundaries() {
+        let s = "🌑🌒🌓🌔";
+        let individually = ["🌑", "🌒", "🌓", "🌔"];
+
+        let mut expect = Vec::new();
+        // [(0, 0, "", ""),
+        //  (0, 4, "🌑", "🌑"), repeated 4 times,
+        //  (4, 4, "", ""), once
+        //  (4, 8, "🌒", "🌒"), repeated 4 times, and so on (+4, +4)
+        //  ...
+        //  (16, 16, "", ""); 18]
+        expect.push((0, 0, "", ""));
+        for (idx, &char) in individually.iter().enumerate() {
+            let n = char.len();
+            assert_eq!(n, 4);
+            let expected_start = (idx % n) * n;
+            let expected_end = (idx % n) * n + n;
+            expect.extend(repeat((expected_start, expected_end, char, char)).take(n - 1));
+            expect.push((expected_end, expected_end, "", ""));
+        }
```

Comment:
We could potentially use property testing here, but I think this is fine for now!

Reply:
I've look at property testing when I get a chance, and keep it in mind.